pkg/resource/deploy/step_generator.go

```diff
@@ -333,175 +334,224 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 	//  It must be an update or a replace. Which operation we do depends on the the specific change made to the
 	//  resource's properties:
+	//
+	//  - if the user has requested that only specific resources be updated, and this resource is
+	//    not in that set, do no 'Diff' and just treat the resource as 'same' (i.e. unchanged).
+	//
 	//  - If the resource's provider reference changed, the resource must be replaced. This behavior is founded upon
 	//    the assumption that providers are recreated iff their configuration changed in such a way that they are no
 	//    longer able to manage existing resources.
+	//
 	//  - Otherwise, we invoke the resource's provider's `Diff` method. If this method indicates that the resource must
 	//    be replaced, we do so. If it does not, we update the resource in place.
 	if hasOld {
 		contract.Assert(old != nil)
-		diff, err := sg.diff(urn, old, new, oldInputs, oldOutputs, inputs, prov, allowUnknowns, goal.IgnoreChanges)
-		if err != nil {
-			// If the plugin indicated that the diff is unavailable, assume that the resource will be updated and
-			// report the message contained in the error.
-			if _, ok := err.(plugin.DiffUnavailableError); ok {
-				diff = plugin.DiffResult{Changes: plugin.DiffSome}
-				sg.plan.ctx.Diag.Warningf(diag.RawMessage(urn, err.Error()))
-			} else {
-				return nil, result.FromError(err)
+
+		// If the user requested only specific resources to update, and this resource was not in
+		// that set, then do nothin but create a SameStep for it.
+		if updateTargetsOpt != nil && !updateTargetsOpt[urn] {
+			logging.V(7).Infof(
+				"Planner decided not to update '%v' due to not being in target group (same) (inputs=%v)", urn, new.Inputs)
+		} else {
+			updateSteps, res := sg.generateStepsFromDiff(
+				event, urn, old, new, oldInputs, oldOutputs, inputs, prov, goal)
+
+			if res != nil {
+				return nil, res
 			}
+
+			if len(updateSteps) > 0 {
+				// 'Diff' produced update steps.  We're done at this point.
+				return updateSteps, nil
+			}
+
+			// Diff didn't produce any steps for this resource.  Fall through and indicate that it
+			// is same/unchanged.
+			logging.V(7).Infof("Planner decided not to update '%v' after diff (same) (inputs=%v)", urn, new.Inputs)
 		}
 
-		// Ensure that we received a sensible response.
-		if diff.Changes != plugin.DiffNone && diff.Changes != plugin.DiffSome {
-			return nil, result.Errorf(
-				"unrecognized diff state for %s: %d", urn, diff.Changes)
+		// No need to update anything, the properties didn't change.
+		sg.sames[urn] = true
+		return []Step{NewSameStep(sg.plan, event, old, new)}, nil
+	}
+
+	if updateTargetsOpt != nil {
+		// The user asked to only update specific targets.  But they ran an app that produced new
+		// resources.  Error and bail out in this case.
+		//
+		// Note: we could attempt to support this in the future by allowing them to explicitly
```

Comment:
Seems like we should allow a `CreateStep` for a resource in the `updateTargetsOpt` set: doing so will allow users that hit this case to unblock themselves.

Reply:
Note: i avoided this initially because it becomes much harder to validate the set of targets specified by the user.  i.e. today we cna immediately check the targets against the previous set of resources, ensuring that every target mentioned is an existing resource (and giving a good error if it isn't).  Now, we would likely need to *not* do that initially for updates, in case someone was mentioning a resource that would be created.  Then allow the work to happen, then error at the end if they mention a resource that either wasn't htere before, or wasn't created.

Reply:
How about emitting a warning in that case at the end of the preview (and perhaps an error at the end of the update)?

Reply:
doing that now.